src/models/discriminant_analysis.jl

```diff
@@ -135,329 +128,366 @@ metadata_model(
+  method::Symbol = :gevd::(_ in (:gevd, :whiten))
+  cov_w::CovarianceEstimator=MS.SimpleCovariance()
+  cov_b::CovarianceEstimator=MS.SimpleCovariance()
+  outdim::Int=0::(_ ≥ 0)
+  regcoef::Float64=1e-6::(_ ≥ 0)
+  priors::Union{
+    Nothing, 
+    UnivariateFinite{<:Any, <:Any, <:Any, <:Real},
+    Dict{<:Any, <:Real}
+  }=nothing
 end
 
-function MMI.fit(model::BayesianLDA, ::Int, X, y)
-    Xm_t, yplain, classes_seen, p, n, nc, nclasses, integers_seen, outdim =
-        _check_lda_data(model, X, y)
-    ## If piors are specified check if they makes sense.
-    ## This was put here to through errors much earlier
-    if isa(model.priors, Vector)
-        priors = _check_lda_priors(model.priors, nc, nclasses, integers_seen)
-    end
-
-    core_res = MS.fit(
-        MS.MulticlassLDA, nc, Xm_t, Int.(yplain);
-        method=model.method,
-        outdim,
-        regcoef=model.regcoef,
-        covestimator_within=model.cov_w,
-        covestimator_between=model.cov_b
-    )
-
-    ## Estimates prior probabilities if specified by user.
-    ## Put it here to avoid recomputing as the fitting process does this already
-    if isa(model.priors, Nothing)
-        weights = MS.classweights(core_res)
-        total = core_res.stats.tweight
-        priors = weights ./ total
-    end
-    cache = nothing
-    report = (
-        classes=classes_seen,
-        outdim=MS.size(core_res)[2],
-        projected_class_means=MS.classmeans(core_res),
-        mean=MS.mean(core_res),
-        class_weights=MS.classweights(core_res),
-        Sw=MS.withclass_scatter(core_res),
-        Sb=MS.betweenclass_scatter(core_res),
-        nclasses=nc
-    )
-
-    fitresult = (core_res, classes_seen, priors, n)
-    return fitresult, cache, report
+function _matrix_transpose(::Union{LDA,BayesianLDA}, X)
+  # MultivariateStats 9.0 is not supporting adjoints
+  return MMI.matrix(X, transpose=true)
 end
 
-function _matrix_transpose(::Union{LDA,BayesianLDA}, X)
-    # MultivariateStats 9.0 is not supporting adjoints
-    return MMI.matrix(X, transpose=true)
+function _check_prob01(priors)
+  # `priors` is esssentially always an instance of type `Vector{<:Real}`.
+  # The next two conditions implicitly checks that
+  # ` 0 .<= priors .<= 1` and `sum(priors) ≈ 1` are true.
+  if any(<(0), priors)
+    throw(ArgumentError("probabilities specified in `priors` must non-negative"))
+  end
+
+  if !isapprox(sum(priors), 1)
+    throw(ArgumentError("probabilities specified in `priors` must sum to 1"))
+  end
+  return nothing
+end
+
+@inline function _check_lda_priors(priors::UnivariateFinite, classes_seen, pool)
+  if MMI.classes(priors) != pool
+      throw(
+        ArgumentError("UnivariateFinite `priors` must have common pool with training target.")
+      )
+  end
+  priors_ = pdf.(priors, support(priors))
+  _check_prob01(priors_)
+  # Select priors for unique classes in `y` (For resampling purporses).
+  return pdf.(priors, classes_seen)
```

Comment:
I thought we were going to make this optional checks on the CategoricalDistributions.jl side. That sounded like a sensible plan.

Reply:
Yes, But on giving it a second thought, we would still need to make checks again here, because there is no way of telling if the `UnivariateFinite` object the user provided has been checked during construction.

Reply:
Aha. I see.